### ISSUE TITLE
ICC metadata gets "lost" when reading an image without an embedded profile

### ISSUE THREAD

**Issue by aferrero2707** [jcupitt/libvips#251]
I have some issues with ICC profiles (and other metadata) added "by hand" to existing VipsImage objects.

I have put a simple test case in this gist: https://gist.github.com/aferrero2707/a33de898d2620f41e5cb (it requires LCMS2).

You can run the example like that: 
    icc_test input_image.jpg output_image.jpg

The example opens an image, sets by hand the VIPS_META_ICC blob to a valid AdobeRGB profile, then it writes back the image to disk.

If you pass an input image WITHOUT an embedded ICC profile, the output is as follows:
    Embedded profile not found after vips_image_new_from_file()
    Embedded profile found after vips_image_set_blob()
    Embedded profile not found after vips_image_write_to_file()

In other words, the operation of writing the image to disk REMOVES the ICC blob from the VipsImage, which is something I would not expect.

The blob gets "erased" by this call in vips_foreign_load_start():

vips_image_pipelinev( load->out, load->out->dhint, load->real, NULL );

Is this expected?

**Comment by aferrero2707** [comment]
Hi!

I think I've pinned down the reason of the behavior I'm observing. Here is what happens, step-by-step:

1. I load an image WITHOUT any embedded ICC profile, using **vips_image_new_from_file()**
2. I create a valid AdobeRGB profile and I put it into the loaded image with 
    **vips_image_set_blob( image, VIPS_META_ICC_NAME,
       (VipsCallbackFn) g_free, buf, out_length );**
    where "buf" is the memory buffer holding the ICC profile
3. I save the image to disk, with 
    **vips_image_write_to_file( image, argv[2], NULL );**

At this point, the ICC metadata has disappeared from "image"...
In fact, **vips_foreign_load_start()** contains a call to

    **vips_image_pipelinev( load->out, load->out->dhint, load->real, NULL );**

where "load->out" is the image containing the ICC metadata and "load->real" is an image created internally from the same disk file and therefore NOT containing any ICC metadata.

**vips_image_pipelinev()** in turn calls **vips__image_copy_fields_array()** and copies the metadata from "load->real" to "load->out".

But while doing that, the original metadata in "load->out" is erased by a call to 

    **vips__meta_destroy( out );**


Due to that, there is no way to preserve the metadata added to an image AFTER it has been loaded from disk, and which is not present in the disk file itself.

However, I'm afraid that simply removing the call to **vips__meta_destroy( out )** in **vips__image_copy_fields_array()** can potentially break the rest of the VIPS library...

**Comment by jcupitt** [comment]
I'm sorry I've been so slow on this issue. I'll try to look into it this week.

**Comment by aferrero2707** [comment]
Don't worry, I've found a temporary workaround for that (although quite an ugly one), also my issue probably does not involve many other users of the library... 

Maybe you could directly jump to my pull request and see if it makes sense: https://github.com/jcupitt/libvips/pull/265

Thanks for looking!

**Comment by aferrero2707** [comment]
An update on this issue: I've found a reasonable good fix (not just a workaround) for the problem of custom metadata fields added to images created via *vips_image_new_from_file()*.

The fix is actually very simple: duplicate the image with *vips_copy()* and then add the metadata to this copy instead of the initial VipsImage. This prevents the metadata to get erased by *vips_foreign_load_start()*.

I think we can close the issue.

**Comment by jcupitt** [comment]
A bit late here, but closing.
